CP-18271 first cut at C library for RRD plugins

lindig opened the pull request:
I'd like to invite comments on this first cut to implement a C library that writes the RRD V2 protocol to a file to be consumed by RRDD.  The README (not part of this PR but in the repository) provides a design overview.

The commit history is messy as I used it just for snapshots; it will be squashed later.

sharady committed:
Initial commit

sharady committed:
Adding header file for rrd generation in C

Signed-off-by: Sharad Yadav <sharad.yadav@citrix.com>

lindig committed:
CP-17654 add Makefile

The Makefile is an early start to compile the code. It reveals that
there are several syntax problems in the code.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 more idiomatic header file

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 add proposal for interface

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 fix Makefile (rdd->rrd), whitespace

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 Fix dependency in Makefile

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 Improve interface, account for value polymorphism

A value can be either a float or an integer. The interface is now using
a union to represent them. Added a README.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 sharpen type definitions: use int64_t; sample(void)

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 add JSON library parson as a submodule; add rrdtest

Signed-off-by: Christian Lindig <lindig@gmail.com>

lindig committed:
Merge branch 'proposal' of github.com:lindig/C-library-for-RRD-generation into proposal

lindig committed:
CP-17654 fix typo in librrd.h (RDD_OK -> RRD_OK)

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 fix type in (private) type RRD_PLUGIN

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 laying groundwork for implementation and test - work in progress

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-17654 generating JSON - work in progress

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 generate JSON for a plugin (work in progress)

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix memory leaks, add valgrind target to Makefile

This is work in progress. Fixed two memory leaks.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 implement buffer initialisation

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 writing checksums to buffer - work in progress

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 update README, htonll()

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 improve documentation, error handling

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 improve documentation, error handling

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 add .gitignore

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
Add LICENSE

lindig committed:
CP-18271 fix Makefile (PHONY targets)

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 make compile on OSX

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 revert change from OS X compatibility

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 document Parson sub module

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
Merge remote-tracking branch 'github-xapi/master'

robhoes commented on librrd.h:
```diff
@@ -0,0 +1,128 @@
+
+#define RRD_OK                  0
+#define RRD_TOO_MANY_SOURCES    1
+#define RRD_NO_SOUCH_SOURCE     2
```
Souch?

lindig committed:
CP-18271 add .travis.yml

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix typo in error constant

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix to get Travis CI to work; use gcc -std=gnu99

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 add simple RRD plugin "rrdclient"

The client reads integers from stdin and reports them.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 add example for a simple client, update README for it

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 add Travis build status

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix typo in URL to Travis CI

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

Zhengchai commented:
Thanks. What is the interface output difference comparing with the current one?

Asking this because of many other components like WLB who are using the RRD output also, keep the same output formats and data will not cause regression.

Also exercise with a full WLB tests and understand the results will be great.

lindig commented:
I don't understand the question. The file format for communicating with RRDD is specified here: https://info.citrite.net/display/engp/RRD+plugin+protocol+V2 and this is what the library aims to implement.

Zhengchai commented:
OK, in a simple word, with V2 RRD, is there any change of existing RRD RPC exposing/Providing data structure and parameters? e.g. When Client requesting data from host with command "HTTP://x.x.x.x:80/rrd_updates?session_id=OpaqueRef:xxxxxxxx&cf=AVERAGE&start=1470036782&host=true", will the output be changed?

I can post your an example of current result if you need to comparing (e.g. with email).

lindig commented:
I don't think this is relevant here. This protocol and library is about data sources reporting data to RRDD and not about how RRDD presents the information it has collected to the outside. So I don't expect any change to current behaviour.

Zhengchai commented:
That will be great if no impacts/changes the output side. just want to ensure this given that kind of change may impact a lot to WLB indeed. thanks for confirming. Let me know please if that is not the case. thanks.

lindig committed:
CP-18271 rewrote code that searches for slots in plugin

This is in response to warnings from a Coverity analysis. The analysis
claims a buffer overrun in the JSON library

  verify_utf8_sequence(const unsigned char *string, int *len)

bit this is a false positive: the analysis assumes that characters in a
string have a multi-byte representation when in fact they don't.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix Makefile for Coverity analysis; check return value in demo

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 made code more robust wrt compiler

Many small changes to make the code more robust against differences
between compilers after review by Andrew Cooper. Improved error
handling, better use of naming conventions.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 update README after latest code changes

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 add OCaml tool to read rrd files for testing

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

robhoes commented:
I think it's great! :)
:shipit:

lindig commented:
I've started testing it with rrd-transport and there are some bugs that I need to weed out first in the data format.

lindig committed:
CP-18271 fix bugs found during test with rrdreader/rrd-transport

* rrd-transport expects the default value to be a string, not a boolean
* the outer JSON object was missing
* test code did not explicitly set the type of the data sources

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix bug in test code: set type of data source

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
improve portability of testing code

Use size designators from inttypes.h in printf.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
avoid re-defining typedef RRD_PLUGIN

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
update .gitignore

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

lindig committed:
CP-18271 fix 64-bit constant used for initialisation of data slot

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>
